finetune/lora.py

```diff
@@ -0,0 +1,225 @@
+    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    out_dir: Path = Path("out/lora/alpaca"),
+    precision: Literal["bf16-true", "32-true", "bf16-mixed"] = "bf16-true",
+):
+
+    print(checkpoint_dir)
+    check_valid_checkpoint_dir(checkpoint_dir)
+
+    fabric = L.Fabric(accelerator="cuda", devices=1, precision=precision)
+    fabric.launch()
+    fabric.seed_everything(1337 + fabric.global_rank)
+
+    if fabric.global_rank == 0:
+        os.makedirs(out_dir, exist_ok=True)
+
+    train_data, val_data = load_datasets(data_dir=data_dir)
+
+    config = Config.from_name(name=checkpoint_dir.name, block_size=max_seq_length)
+    config.block_size = max_seq_length
+
+    checkpoint_path = checkpoint_dir / "lit_model.pth"
+   
+    fabric.print(f"Loading model {str(checkpoint_path)!r} with {config.__dict__}")
+    
+    #with fabric.init_module():
+    #    model = Parrot(config)
+    #with lazy_load(checkpoint_dir / "lit_model.pth") as checkpoint:
+    #    # strict=False because missing keys due to adapter weights not contained in state dict
+    #    model.load_state_dict(checkpoint, strict=False)
+
+    with fabric.init_module(), lora(r=lora_r, alpha=lora_alpha, dropout=lora_dropout, enabled=True):
+        model = Parrot(config)
+        # strict=False because missing keys due to LoRA weights not contained in checkpoint state
+        model.load_state_dict(torch.load(checkpoint_path), strict=False)
+
+    mark_only_lora_as_trainable(model)
+    num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
+    fabric.print(f"Number of trainable parameters: {num_params}")
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
+    model, optimizer = fabric.setup(model, optimizer)
+    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir)
+
+    # Save the final LoRA checkpoint at the end of training
+    checkpoint = lora_state_dict(model)
+
+    fabric.save(os.path.join(out_dir, "lit-model-lora-finetuned.pth"), checkpoint)
+
+
+def train(
+    fabric: L.Fabric,
+    model: torch.nn.Module,
+    optimizer: torch.optim.Optimizer,
+    train_data: np.ndarray,
+    val_data: np.ndarray,
+    checkpoint_dir: Path,
+    out_dir: str,
+) -> None:
+    """The training loop.
+
+    Loosely based on the nanoGPT implementation: https://github.com/karpathy/nanoGPT.
+    """
+
+    step_count = 0
+
+    tokenizer = Tokenizer(checkpoint_dir / "tokenizer.json", checkpoint_dir / "tokenizer_config.json")
+
+    for iter_num in range(max_iters):
+
+        if step_count <= warmup_iters:
+            # linear warmup
+            lr = learning_rate * step_count / warmup_iters
+            for param_group in optimizer.param_groups:
+                param_group['lr'] = lr
+
+        t0 = time.time()
+
+        input_ids, targets = get_batch(fabric, train_data)
+        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_iters != 0)):
```

Comment:
Can you incorporate the changes from #110?

Reply:
done